File: AndroidCover/app/src/main/java/com/example/shakil/androidcover/MainActivity.java
package com.example.shakil.androidcover;

import android.app.AlertDialog;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.AdapterView;
import android.widget.TextSwitcher;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ViewSwitcher;

import com.example.shakil.androidcover.Adapter.MovieAdapter;
import com.example.shakil.androidcover.Common.Common;
import com.example.shakil.androidcover.Interface.IFirebaseLoadDone;
import com.example.shakil.androidcover.Model.Movie;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

import dmax.dialog.SpotsDialog;
import it.moondroid.coverflow.components.ui.containers.FeatureCoverFlow;

public class MainActivity extends AppCompatActivity implements IFirebaseLoadDone {

    IFirebaseLoadDone iFirebaseLoadDone;
    DatabaseReference dbRef;

    AlertDialog dialog;

    //View
    FeatureCoverFlow coverFlow;
    TextSwitcher title;

    MovieAdapter movieAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setContentView(R.layout.activity_main);

        iFirebaseLoadDone = this;
        loadData();
    }

    private void loadData() {
        //Create dialog
        dialog = new SpotsDialog.Builder().setContext(MainActivity.this)
                .setMessage("Please wait...")
                .setCancelable(false)
                .build();
        dialog.show();

        dbRef = FirebaseDatabase.getInstance().getReference("Movies");
        dbRef.addListenerForSingleValueEvent(new ValueEventListener() {
            List<Movie> movies = new ArrayList<>();
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                for (DataSnapshot movieSnapShot : dataSnapshot.getChildren()){
                    Movie movie = movieSnapShot.getValue(Movie.class);
                    movies.add(movie);
                }
                iFirebaseLoadDone.onFirebaseLoadSuccess(movies);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                iFirebaseLoadDone.onFirebaseLoadFailed(databaseError.getMessage());
            }
        });
    }

    @Override
    public void onFirebaseLoadSuccess(List<Movie> movieList) {
        dialog.dismiss();
        setupUI();
        Common.movieLoaded = movieList;

        movieAdapter = new MovieAdapter(this, movieList);
        coverFlow.setAdapter(movieAdapter);
        coverFlow.setOnScrollPositionListener(new FeatureCoverFlow.OnScrollPositionListener() {
            @Override
            public void onScrolledToPosition(int position) {
                title.setText(Common.movieLoaded.get(position).getName());
            }

            @Override
            public void onScrolling() {

            }
        });
        coverFlow.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Intent intent = new Intent(MainActivity.this, DetailActivity.class);
                intent.putExtra("movie_index", position);
                startActivity(intent);
            }
        });
    }

    private void setupUI() {
        setContentView(R.layout.activity_main);
        coverFlow = findViewById(R.id.coverFlow);
        title = findViewById(R.id.title);
        title.setFactory(new ViewSwitcher.ViewFactory() {
            @Override
            public View makeView() {
                LayoutInflater inflater = LayoutInflater.from(MainActivity.this);
                TextView txt = (TextView) inflater.inflate(R.layout.layout_title, null);
                return txt;
            }
        });

        Animation in = AnimationUtils.loadAnimation(this, android.R.anim.slide_in_left);
        Animation out = AnimationUtils.loadAnimation(this, android.R.anim.slide_out_right);
        title.setInAnimation(in);
        title.setOutAnimation(out);

        //Calculate screen size
        Display display = getWindowManager().getDefaultDisplay();
        DisplayMetrics outMetrics = new DisplayMetrics();
        display.getMetrics(outMetrics);

        float density = getResources().getDisplayMetrics().density;
        float dpHeight = outMetrics.heightPixels/density;
        float dpWidth = outMetrics.widthPixels/density;

        coverFlow.setCoverHeight((int) dpHeight);
        coverFlow.setCoverWidth((int) dpWidth);

    }

    @Override
    public void onFirebaseLoadFailed(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }
}

File: KotlinCover/app/src/main/java/com/example/shakil/kotlincover/MainActivity.kt
package com.example.shakil.kotlincover

import android.app.AlertDialog
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.animation.AnimationUtils
import android.widget.AdapterView
import android.widget.TextView
import android.widget.Toast
import android.widget.ViewSwitcher
import com.example.shakil.kotlincover.Adapter.MovieAdapter
import com.example.shakil.kotlincover.Common.Common
import com.example.shakil.kotlincover.Interface.IFirebaseLoadDone
import com.example.shakil.kotlincover.Model.Movie
import com.google.firebase.database.*
import dmax.dialog.SpotsDialog
import it.moondroid.coverflow.components.ui.containers.FeatureCoverFlow
import kotlinx.android.synthetic.main.activity_main.*
import java.util.ArrayList

class MainActivity : AppCompatActivity(), IFirebaseLoadDone {

    lateinit var iFirebaseLoadDone: IFirebaseLoadDone
    lateinit var dbRef:DatabaseReference

    lateinit var dialog: AlertDialog
    lateinit var adapter: MovieAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //setContentView(R.layout.activity_main)

        iFirebaseLoadDone = this
        loadData()
    }

    private fun loadData() {
        //Create dialog
        dialog = SpotsDialog.Builder().setContext(this@MainActivity)
            .setMessage("Please wait...")
            .setCancelable(false)
            .build()
        dialog.show()

        dbRef = FirebaseDatabase.getInstance().getReference("Movies")
        dbRef.addListenerForSingleValueEvent(object : ValueEventListener {

            var movies: MutableList<Movie> = ArrayList()

            override fun onDataChange(dataSnapshot: DataSnapshot) {
                for (movieSnapShot in dataSnapshot.children) {
                    val movie = movieSnapShot.getValue(Movie::class.java)
                    movies.add(movie!!)
                }
                iFirebaseLoadDone.onFirebaseLoadSuccess(movies)
            }

            override fun onCancelled(databaseError: DatabaseError) {
                iFirebaseLoadDone.onFirebaseLoadFailed(databaseError.message)
            }
        })
    }

    override fun onFirebaseLoadSuccess(movieList: List<Movie>) {
        dialog.dismiss()
        setupUI()
        Common.movieLoaded = movieList

        adapter = MovieAdapter(this, movieList)
        cover_flow.adapter = adapter
        cover_flow.setOnScrollPositionListener(object : FeatureCoverFlow.OnScrollPositionListener {
            override fun onScrolledToPosition(position: Int) {
                textSwitcher.setText(Common.movieLoaded[position].name)
            }

            override fun onScrolling() {

            }
        })
        cover_flow.onItemClickListener = AdapterView.OnItemClickListener { adapterView, view, position, id ->
            val intent = Intent(this@MainActivity, DetailActivity::class.java)
            intent.putExtra("movie_index", position)
            startActivity(intent)
        }
    }

    private fun setupUI() {
        setContentView(R.layout.activity_main)

        textSwitcher.setFactory {
            val inflater = LayoutInflater.from(this@MainActivity)
            inflater.inflate(R.layout.layout_title, null) as TextView
        }

        val `in` = AnimationUtils.loadAnimation(this, android.R.anim.slide_in_left)
        val out = AnimationUtils.loadAnimation(this, android.R.anim.slide_out_right)
        textSwitcher.inAnimation = `in`
        textSwitcher.outAnimation = out

        //Calculate screen size
        val display = windowManager.defaultDisplay
        val outMetrics = DisplayMetrics()
        display.getMetrics(outMetrics)

        val density = resources.displayMetrics.density
        val dpHeight = outMetrics.heightPixels / density
        val dpWidth = outMetrics.widthPixels / density

        cover_flow.coverHeight = dpHeight.toInt()
        cover_flow.coverWidth = dpWidth.toInt()
    }

    override fun onFirebaseLoadFailed(message: String) {
        Toast.makeText(this@MainActivity, message, Toast.LENGTH_SHORT).show()
    }
}
